File: ide-features-trainer/src/org/jetbrains/training/commands/TypeTextCommand.java
package org.jetbrains.training.commands;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import org.jdom.Element;
import org.jetbrains.training.commandsEx.CommandEx;
import org.jetbrains.training.editor.MouseListenerHolder;
import org.jetbrains.training.graphics.DetailPanel;
import org.jetbrains.training.lesson.Lesson;

import java.util.Queue;
import java.util.concurrent.ExecutionException;

/**
 * Created by karashevich on 30/01/15.
 */
public class TypeTextCommand extends Command {

    public TypeTextCommand(){
        super(CommandType.TYPETEXT);
    }

    @Override
    public void execute(final ExecutionList executionList) throws InterruptedException, ExecutionException {

        Element element = executionList.getElements().poll();

        final String finalText = (element.getContent().isEmpty() ? "" : element.getContent().get(0).getValue());
        boolean isTyping = true;
        final int[] i = {0};
        final int initialOffset = executionList.getEditor().getCaretModel().getOffset();

        while (isTyping) {
//            Thread.sleep(20);
            final int finalI = i[0];
            WriteCommandAction.runWriteCommandAction(executionList.getProject(), new Runnable() {
                @Override
                public void run() {
                    executionList.getEditor().getDocument().insertString(finalI + initialOffset, finalText.subSequence(i[0], i[0] + 1));
                    executionList.getEditor().getCaretModel().moveToOffset(finalI + 1 + initialOffset);
                }
            });
            isTyping = (++i[0] < finalText.length());
        }

        //execute next
        startNextCommand(executionList);

    }
}

File: ide-features-trainer/src/training/commands/TypeTextCommand.kt
/*
 * Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */
package training.commands

import com.intellij.openapi.command.WriteCommandAction

class TypeTextCommand : Command(CommandType.TYPETEXT) {

  override fun execute(executionList: ExecutionList) {
    executeWithPollOnEdt(executionList) {
      element ->

      val textToType = if (element.content.isEmpty()) "" else element.content[0].value
      val startOffset = executionList.editor.caretModel.offset

      var isTyping = true
      val i = intArrayOf(0)

      while (isTyping) {
        val finalI = i[0]
        WriteCommandAction.runWriteCommandAction(executionList.project, Runnable {
          executionList.editor.document.insertString(finalI + startOffset, textToType.subSequence(i[0], i[0] + 1))
          executionList.editor.caretModel.moveToOffset(finalI + 1 + startOffset)
        })
        isTyping = ++i[0] < textToType.length
      }
    }
  }

}
